src/hardware/signal_generator.rs

```diff
@@ -0,0 +1,176 @@
+use crate::hardware::{dac::DacCode, design_parameters::ADC_SAMPLE_TICKS};
+use core::convert::{TryFrom, TryInto};
+use miniconf::Miniconf;
+use serde::Deserialize;
+
+/// Types of signals that can be generated.
+#[derive(Copy, Clone, Debug, Deserialize, Miniconf)]
+pub enum Signal {
+    Cosine,
+    Square,
+    Triangle,
+}
+
+/// Basic configuration for a generated signal.
+///
+/// # Miniconf
+/// `{"signal": <signal>, "frequency", 1000.0, "symmetry": 0.5, "amplitude": 1.0}`
+///
+/// Where `<signal>` may be any of [Signal] variants, `frequency` specifies the signal frequency
+/// in Hertz, `symmetry` specifies the normalized signal symmetry which ranges from 0 - 1.0, and
+/// `amplitude` specifies the signal amplitude in Volts.
+#[derive(Copy, Clone, Debug, Miniconf, Deserialize)]
+pub struct BasicConfig {
+    /// The signal type that should be generated. See [Signal] variants.
+    pub signal: Signal,
+
+    /// The frequency of the generated signal in Hertz.
+    pub frequency: f32,
+
+    /// The normalized symmetry of the signal. At 0% symmetry, the first half phase does not exist.
+    /// At 25% symmetry, the first half-phase lasts for 25% of the signal period.
+    pub symmetry: f32,
```

Comment:
Disambiguating the definition of `phase`:

```suggestion
    /// The normalized symmetry of the signal. At 0% symmetry, the duration of the first half oscillation is minimal.
    /// At 25% symmetry, the first half oscillation lasts for 25% of the signal period. For square wave output this
    //// symmetry is the duty cycle.
    pub symmetry: f32,
```